### PR TITLE
demo: add GMAC demonstration program

### LINKED ISSUE #14123: Add an EVP demo for GMAC
It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by paulidale**
Test vector from NIST's GCM mode suite.

Fixes #14123

### CODE REVIEW

File: demos/mac/gmac.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    OSSL_LIB_CTX *library_context = NULL;
+    size_t out_len = 0;
+
+    library_context = OSSL_LIB_CTX_new();
+    if (library_context == NULL) {
+        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /* Fetch the key derivation function implementation */
```
Comment (slontis) [inline demos/mac/gmac.c:74]: KDF?

File: demos/mac/gmac.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    mac = EVP_MAC_fetch(library_context, "GMAC", propq);
+    if (mac == NULL) {
+        fprintf(stderr, "EVP_MAC_fetch() returned NULL\n");
+        goto end;
+    }
+
+    /* Create a context for the key derivation operation */
```
Comment (slontis) [inline demos/mac/gmac.c:81]: And here also.

File: demos/mac/gmac.c (hunk 1 above)
Comment (paulidale) [reply demos/mac/gmac.c:74]: Oops :)

File: demos/mac/gmac.c (hunk 2 above)
Comment (paulidale) [reply demos/mac/gmac.c]: Fix pushed.

**Comment by paulidale** [comment]
Ping for review.

File: demos/mac/gmac.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    if (CRYPTO_memcmp(expected_output, out, sizeof(expected_output)) != 0) {
+        fprintf(stderr, "Generated MAC does not match expected value\n");
+        goto end;
+    }
+
+    rv = EXIT_SUCCESS;
+end:
+    EVP_MAC_CTX_free(mctx);
```
Comment (slontis) [inline demos/mac/gmac.c:135]: Add...
    if (ret == EXIT_FAILURE)
        ERR_print_errors_fp(stderr);

File: demos/mac/gmac.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        fprintf(stderr, "EVP_MAC_update() failed\n");
+        goto end;
+    }
+
+    /* Make one call to the final to get the MAC */
+    if (!EVP_MAC_final(mctx, out, &out_len, sizeof(out))) {
+        fprintf(stderr, "EVP_MAC_final() failed\n");
+        goto end;
+    }
+
```
Comment (slontis) [inline demos/mac/gmac.c:122]: Maybe dumping the output is also nice..
    fprintf(stdout, "Generated MAC:\n");
    BIO_dump_indent_fp(stdout, out, out_len, 2);
    fprintf(stdout, "\n");

File: demos/mac/gmac.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                                             iv, sizeof(iv));
+    *p = OSSL_PARAM_construct_end();
+
+    /* Initialise the GMAC operation */
+    if (!EVP_MAC_init(mctx, key, sizeof(key), params)) {
+        fprintf(stderr, "EVP_MAC_init() failed\n");
+        goto end;
+    }
+
+    /* Make one or more calls to process the data to be authenticated */
```
Comment (slontis) [inline demos/mac/gmac.c:111]: Is this comment right? It generates an auth tag..

File: demos/mac/gmac.c (hunk 4 above)
Comment (paulidale) [reply demos/mac/gmac.c:122]: Added.

File: demos/mac/gmac.c (hunk 3 above)
Comment (paulidale) [reply demos/mac/gmac.c:139]: Added

File: demos/mac/gmac.c (hunk 5 above)
Comment (paulidale) [reply demos/mac/gmac.c:111]: I think it's right.  The data needs to be processed before the tag can be obtained.  I.e. we're allowing the data to be authenticated later.

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Merged with a whitespace correction.  Thanks for the feedback.

**Pull request closed without merging**